Fix superprimes and superwavemechanics so pan works

bgold-cosmos opened the pull request.

telephon committed:
Merge pull request #112 from musikinformatik/develop

Sync with develop

telephon committed:
Merge pull request #116 from musikinformatik/develop

dirt-library: do not confuse numChannels with buffer.numChannels

bgold-cosmos committed:
Merge branch 'develop' of https://github.com/musikinformatik/SuperDirt

bgold-cosmos committed:
Fix multidimensional arrays so pan works properly

telephon reviewed (commented):
are you sure you want to use a stereo mix to then pan to numChannels? Also, the `scramble` only happens when the synthdef is built, that is when you completely start supercollider. Maybe there is a better solution round the corner?

telephon commented:
@bgold-cosmos any follow up? It would be good to either merge or close soon.

bgold-cosmos commented:
Whoops, yeah, got sidetracked thinking about multichannel expansions.  I'm traveling the next few days but I'll put together something soon.

bgold-cosmos committed:
Update default-synths-extra.scd

Let DirtPan do the multichannel mixdown, and use continuous randomness rather than just initial

bgold-cosmos commented:
I think this is much better although at the moment I don't have the setup to thoroughly test multichannel output.

telephon commented on library/default-synths-extra.scd:
```diff
@@ -480,13 +481,13 @@ SynthDef(\superwavemechanics, {|out, pan, freq, sustain, accelerate, detune=0, v
 	freq = freq * XLine.ar(1, exp(accelerate), sustain);
 	env = EnvGen.ar(Env.sine(1.0), timeScale:sustain, doneAction:2);
-	i = (0..7).scramble;
+	i = Array.fill(8, {TIRand.kr(0, 3, Impulse.kr(8))});
```
why trigger at 8 Hz? Also, if `i` is used in line 487 in a `Rand` it will never be retriggered.

telephon commented on library/default-synths-extra.scd:
```diff
@@ -465,7 +465,8 @@ SynthDef(\superprimes, {|out, pan, freq, sustain, accelerate, speed=1, detune=0,
     	* LPF.ar(Decay.ar(trig, (primes+3)*sustain/10), voice.linexp(0,2,30,18000));
 	sound = GVerb.ar(sound.softclip, 10*voice.linexp(0,5,1,0.01), 10).tanh;
-	OffsetOut.ar(out, DirtPan.ar(0.5*Splay.ar(sound.scramble.flat), ~dirt.numChannels, pan, env));
+	sound = SplayAz.ar(9, sound.flat, center:LFNoise2.kr(sustain).range(0,8));
```
Is there a particular reason why 9 channels?

bgold-cosmos committed:
Update default-synths-extra.scd

tweaks to address comments:
In superprime, constants of "9" have been changed to be the size of the "primes" array
In superwavemechanics, the detune array is now also a triggered random number, and the index array "i" triggers at a rate determined relative to the sustain (8x)

telephon commented on library/default-synths-extra.scd:
```diff
@@ -481,10 +481,10 @@ SynthDef(\superwavemechanics, {|out, pan, freq, sustain, accelerate, detune=0, v
 	freq = freq * XLine.ar(1, exp(accelerate), sustain);
 	env = EnvGen.ar(Env.sine(1.0), timeScale:sustain, doneAction:2);
-	i = Array.fill(8, {TIRand.kr(0, 3, Impulse.kr(8))});
+	i = Array.fill(8, {TIRand.kr(0, 3, Impulse.kr(sustain/8))});
```
because the 8 `Impulse`Ugens are all the same, you can just have one of them outside the Array and use it as an input for all.

`i = TIRand.kr(0, 3, Impulse.kr(sustain/8).dup(8)))` will expand and do the same.

telephon commented on library/default-synths-extra.scd:
```diff
@@ -481,10 +481,10 @@ SynthDef(\superwavemechanics, {|out, pan, freq, sustain, accelerate, detune=0, v
 	r = [3,5,8,9]/8;
 	detune = min(detune, 1.5);
-	d = Rand(1.6*i - detune, 1.7*i + detune).round(0.25);
+	d = TRand.ar(1.6*i - detune, 1.7*i + detune, Impulse.ar(0)).round(0.25);
```
because you are triggering only once (`Impulse.ar(0)`), you can use a `Rand` instead:

`d = Rand.ar(1.6 * i - detune, 1.7 * i + detune).round(0.25);`

telephon reviewed (changes requested):
yes better!

I've added a few hints for (minor) improvement...

bgold-cosmos committed:
fixes as suggested by @telephon
